chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+            return False
+        with open(filename, 'r') as f:
+            # This is meant to be a best effort attempt.
+            # This can return True and still have no packages
+            # actually being specified, but those aren't common
+            # cases.
+            for line in f:
+                line = line.strip()
+                if line and not line.startswith('#'):
+                    return True
+        return False
+
+    def _download_all_dependencies(self, requirements_file, directory):
+        # type: (str, str) -> set[Package]
+        # Download dependencies prefering wheel files but falling back to
+        # raw source dependences to get the transitive closure over
+        # the dependency graph. Return the set of all package objects
+        # which will serve as the master list of dependencies needed to deploy
+        # successfully.
+        self._pip.download_all_dependencies(requirements_file, directory)
+        deps = {Package(directory, filename) for filename
+                in self._osutils.get_directory_contents(directory)}
+        return deps
+
+    def _download_binary_wheels(self, packages, directory):
+        # type: (set[Package], str) -> None
+        # Try to get binary whls for each package that isn't compatible.
+        self._pip.download_manylinux_whls(
+            [pkg.identifier for pkg in packages], directory)
+
+    def _build_sdists(self, sdists, directory):
+        # type: (set[Package], str) -> None
+        for sdist in sdists:
+            path_to_sdist = self._osutils.joinpath(directory, sdist.filename)
+            self._pip.build_wheel(path_to_sdist, directory)
+
+    def _categorize_whl_files(self, directory):
+        # type: (str) -> Tuple[Set[Package], Set[Package]]
+        final_whls = [Package(directory, filename) for filename
+                      in self._osutils.get_directory_contents(directory)
+                      if filename.endswith('.whl')]
+        valid_whls, invalid_whls = set(), set()
+        for whl in final_whls:
+            if self._valid_lambda_whl(whl.filename):
+                valid_whls.add(whl)
+            else:
+                invalid_whls.add(whl)
+        return valid_whls, invalid_whls
+
+    def _download_dependencies(self, directory, requirements_file):
+        # type: (str, str) -> Tuple[Set[Package], Set[Package]]
+        # Download all dependencies we can, letting pip choose what to download
+        deps = self._download_all_dependencies(requirements_file, directory)
+
+        # Sort the downloaded packages into three categories:
+        # - sdists (Pip could not get a wheel so it gave us a sdist)
+        # - valid whls (lambda compatbile wheel files)
+        # - invalid whls (lambda incompatible wheel files)
+        valid_whls, invalid_whls = self._categorize_whl_files(directory)
+        sdists = deps - valid_whls - invalid_whls
+
+        # Find which packages we do not yet have a valid whl file for. And
```

Comment:
I think I've made this comment before, but rather than explaining what the code is doing, it would really helpful if you explain *why* you're doing this.  As someone not as familiar with the intricacies of pip, a high level overview of why we're doing this multipass approach, why an invalid wheel might be able to be "converted" to a valid manylinux1 wheel, etc.  Putting a more detailed explanation of your algorithm outlined in the PR description would be better in the code itself IMO.